### PR TITLE
Deneb Support

### PR DESCRIPTION
**Pull request by avalonche**
## 📝 Summary

mev-boost-relay changes for the deneb upgrade. See https://github.com/flashbots/mev-boost-relay/issues/489 for changes included in this branch. Summary:

* Clean up of custom types in `common/types.go` and use attestant types instead
* Default to `PublishBlockV2` endpoint for block publishing
* Deneb types and blob validation

## 📚 References

**Specs**
- Builder Specs - https://github.com/ethereum/builder-specs around deneb types
- Dropping blinded blob sidecar - https://github.com/ethereum/builder-specs/pull/90 - short Merkle proof from the signed header's body root to each KZG commitment gossiped by the CL via the `PublishBlockV2` endpoint 
- Relay Specs - https://github.com/flashbots/relay-specs around the deneb block validation endpoint
- Details on the 4844 upgrade - https://www.eip4844.com/.

---

## ✅ I have run these commands

* [x] `make lint`
* [x] `make test-race`
* [x] `go mod tidy`
* [x] I have seen and agree to `CONTRIBUTING.md`

### CODE REVIEW

File: datastore/memcached_test.go [hunk 1]
```diff
@@ -25,50 +35,89 @@ var (
...
+						LogsBloom:     [256]byte{0x05},
+						PrevRandao:    phase0.Hash32{0x06},
+						BlockNumber:   5001,
+						GasLimit:      5002,
+						GasUsed:       5003,
+						Timestamp:     5004,
+						ExtraData:     []byte{0x07},
+						BaseFeePerGas: uint256.NewInt(123),
+						BlockHash:     phase0.Hash32{0x09},
+						Transactions:  []bellatrix.Transaction{},
```
Comment (jtraglia) [inline datastore/memcached_test.go:71]: `BlobGasUsed` and `ExcessBlobGas` are missing values.

File: services/api/types_test.go [hunk 2]
```diff
@@ -3,63 +3,144 @@ package api
...
+							LogsBloom:     [256]byte{0x05},
+							PrevRandao:    phase0.Hash32{0x06},
+							BlockNumber:   5001,
+							GasLimit:      5002,
+							GasUsed:       5003,
+							Timestamp:     5004,
+							ExtraData:     []byte{0x07},
+							BaseFeePerGas: uint256.NewInt(123),
+							BlockHash:     phase0.Hash32{0x09},
+							Transactions:  []bellatrix.Transaction{},
```
Comment (jtraglia) [inline services/api/types_test.go:93]: Also missing new fields here.

File: common/types_spec.go [hunk 3]
```diff
@@ -40,211 +47,393 @@ func BuildGetHeaderResponse(payload *BuilderSubmitBlockRequest, sk *bls.SecretKe
...
+		Blobs:     blockPayload.BlobsBundle.Blobs,
 	}
+}
 
-	return nil, ErrEmptyPayload
+type BuilderBlockValidationRequest struct {
+	*VersionedSubmitBlockRequest
+	RegisteredGasLimit    uint64       `json:"registered_gas_limit,string"`
+	ParentBeaconBlockRoot *phase0.Root `json:"parent_beacon_block_root,omitempty"`
 }
```
Comment (jtraglia) [inline common/types_spec.go:247]: Will the validator function allow `ParentBeaconBlockRoot` to be missing? It doesn't appear to be an optional field in the builder's [`BuilderBlockValidationRequestV3`](https://github.com/flashbots/builder/blob/ae7357b26b3c6bf6e0bb97bf6669e62359854339/eth/block-validation/api.go#L193-L197) struct. Should `BuilderBlockValidationRequest` have the `omitempty` qualifier?

image

File: README.md [hunk 4]
```diff
@@ -152,7 +152,7 @@ redis-cli DEL boost-relay/sepolia:validators-registration boost-relay/sepolia:va
 * `DISABLE_LOWPRIO_BUILDERS` - reject block submissions by low-prio builders
 * `FORCE_GET_HEADER_204` - force 204 as getHeader response
 * `ENABLE_IGNORABLE_VALIDATION_ERRORS` - enable ignorable validation errors
-* `USE_V2_PUBLISH_BLOCK_ENDPOINT` - uses the v2 publish block endpoint on the beacon node
+* `USE_V1_PUBLISH_BLOCK_ENDPOINT` - uses the v1 publish block endpoint on the beacon node
```
Comment (metachris) [inline README.md:155]: will be important to note in the release notes that this switches to v2 endpoint by default

File: beaconclient/multi_beacon_client.go [hunk 5]
```diff
@@ -29,7 +29,11 @@ const (
 )
 
 func (b BroadcastMode) String() string {
-	return [...]string{"gossip", "consensus", "consensus_and_equivocation"}[b]
+	broadcastModeStrings := [...]string{"gossip", "consensus", "consensus_and_equivocation"}
+	if int(b) >= len(broadcastModeStrings) {
+		return "invalid broadcast mode value"
```
Comment (metachris) [inline beaconclient/multi_beacon_client.go:34]: hmmm, not sure this is a good pattern, returning an error as default string as if nothing was wrong... 

i think there's probably a better way to do this whole thing 🤔 like defining the broadcastmodes as string consts?

```go
type BroadcastMode string

const (
    BroadcastModeGossip BroadcastMode = "gossip"
    BroadcastModeConsensus      = "consensus"
    BroadcastModeConsensusAndEquivocation      = "consensus_and_equivocation"
)
```

File: beaconclient/prod_beacon_instance.go [hunk 6]
```diff
@@ -253,32 +253,34 @@ type GetBlockResponse struct {
...
-func (c *ProdBeaconInstance) PublishBlock(block *common.SignedBeaconBlock, broadcastMode BroadcastMode) (code int, err error) {
+func (c *ProdBeaconInstance) PublishBlock(block *common.VersionedSignedProposal, broadcastMode BroadcastMode) (code int, err error) {
 	var uri string
-	if c.ffUseV2PublishBlockEndpoint {
-		uri = fmt.Sprintf("%s/eth/v2/beacon/blocks?broadcast_validation=%s", c.beaconURI, broadcastMode.String())
-	} else {
+	headers := http.Header{}
+	headers.Add("Eth-Consensus-Version", block.Version.String()) // optional in v1, required in v2
+
+	if c.ffUseV1PublishBlockEndpoint {
```
Comment (metachris) [inline beaconclient/prod_beacon_instance.go:277]: i think the pattern before was slightly better, where the if only configures the URL and everything after it is the same

File: beaconclient/util.go [hunk 7]
```diff
@@ -30,28 +31,53 @@ func parseBroadcastModeString(s string) (BroadcastMode, bool) {
...
+	if err != nil {
+		return nil, fmt.Errorf("invalid request for %s: %w", url, err)
+	}
+	// Set content-type
+	req.Header.Add("Content-Type", "application/json")
+	return req, nil
+}
+
+func makeSSZRequest(method, url string, payload any) (*http.Request, error) {
+	if payloadBytes, ok := payload.([]byte); ok {
```
Comment (metachris) [inline beaconclient/util.go:49]: the more typical pattern is checking for `!ok` and returning the error, instead of working in a nested way

File: common/types_spec.go [hunk 8]
```diff
@@ -40,211 +47,393 @@ func BuildGetHeaderResponse(payload *BuilderSubmitBlockRequest, sk *bls.SecretKe
...
+		ParentBeaconBlockRoot *phase0.Root `json:"parent_beacon_block_root,omitempty"`
+	}{
+		RegisteredGasLimit:    r.RegisteredGasLimit,
+		ParentBeaconBlockRoot: r.ParentBeaconBlockRoot,
+	})
 	if err != nil {
 		return nil, err
 	}
+	attrs[0] = ','
+	return append(blockRequest[:len(blockRequest)-1], attrs...), nil
```
[the remark this answers is not in the document]
Comment (metachris) [reply common/types_spec.go:266]: agree, this feels a bit weird and obscure here... where is this coming from, and why? 

it's hard to follow what's happening here.

File: common/types_spec.go [hunk 9]
```diff
@@ -40,211 +47,393 @@ func BuildGetHeaderResponse(payload *BuilderSubmitBlockRequest, sk *bls.SecretKe
...
+func (r *VersionedSubmitBlockRequest) MarshalSSZ() ([]byte, error) {
+	switch r.Version {
+	case spec.DataVersionCapella:
+		return r.Capella.MarshalSSZ()
+	case spec.DataVersionDeneb:
+		return r.Deneb.MarshalSSZ()
+	case spec.DataVersionUnknown, spec.DataVersionPhase0, spec.DataVersionAltair, spec.DataVersionBellatrix:
+		fallthrough
+	default:
+		return nil, errors.Wrap(ErrInvalidVersion, fmt.Sprintf("%d is not supported", r.Version))
```
[the remark this answers is not in the document]
Comment (metachris) [reply common/types_spec.go:282]: good catch!

File: datastore/redis.go [hunk 10]
```diff
@@ -86,6 +83,7 @@ type RedisCache struct {
 	// prefixes (keys generated with a function)
 	prefixGetHeaderResponse           string
 	prefixExecPayloadCapella          string
+	prefixPayloadContentsDeneb        string
```
Comment (metachris) [inline datastore/redis.go:86]: it's okay with me, but why not follow the prev pattern (`prefixExecPayloadCapella` -> `prefixExecPayloadDeneb`)?

File: services/api/service.go [hunk 11]
```diff
@@ -1201,48 +1266,61 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
...
 		return
 	}
 
 	// Take time after the decoding, and add to logging
 	decodeTime := time.Now().UTC()
-	slotStartTimestamp := api.genesisInfo.Data.GenesisTime + (payload.Slot() * common.SecondsPerSlot)
+	slot, err := payload.Slot()
+	if err != nil {
+		log.WithError(err).Warn("failed to get payload slot")
+		api.RespondError(w, http.StatusBadRequest, "failed to get payload slot")
```
Comment (metachris) [inline services/api/service.go:1281]: should return here, after the error?

File: services/api/service.go [hunk 12]
```diff
@@ -1201,48 +1266,61 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
...
+	}
+	blockHash, err := payload.ExecutionBlockHash()
+	if err != nil {
+		log.WithError(err).Warn("failed to get payload block hash")
+		api.RespondError(w, http.StatusBadRequest, "failed to get payload block hash")
+	}
+	proposerIndex, err := payload.ProposerIndex()
+	if err != nil {
+		log.WithError(err).Warn("failed to get payload proposer index")
+		api.RespondError(w, http.StatusBadRequest, "failed to get payload proposer index")
```
Comment (metachris) [inline services/api/service.go:1291]: should return here, after the error?

File: services/api/service.go [hunk 13]
```diff
@@ -1462,10 +1549,24 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
...
 	log = log.WithFields(logrus.Fields{
-		"numTx":       getPayloadResp.NumTx(),
-		"blockNumber": payload.BlockNumber(),
+		"numTx":       len(txs),
+		"blockNumber": blockNumber,
 	})
+	// deneb specific logging
+	if getPayloadResp.Deneb != nil {
+		log = log.WithFields(logrus.Fields{
+			"numBlobs": len(getPayloadResp.Deneb.BlobsBundle.Blobs),
```
Comment (metachris) [inline services/api/service.go:1567]: didn't we also want to log further fields `blob_gas_limit` and `blob_gas_used`?

File: services/api/service.go [hunk 14]
```diff
@@ -1783,30 +1897,38 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
...
 		"isLargeRequest":         isLargeRequest,
 	})
-
-	if payload.Message() == nil || !payload.HasExecutionPayload() {
-		api.RespondError(w, http.StatusBadRequest, "missing parts of the payload")
-		return
+	// deneb specific logging
+	if payload.Deneb != nil {
+		log = log.WithFields(logrus.Fields{
+			"numBlobs": len(payload.Deneb.BlobsBundle.Blobs),
```
Comment (metachris) [inline services/api/service.go:1922]: could also add additional fields - `blob_gas_used` and `blob_gas_limit`?

File: datastore/redis.go (hunk 10 above)
Comment (jtraglia) [reply datastore/redis.go:86]: I can answer this one. The type has changed to include the payload & blobs. `PayloadContents` is accurate.

File: services/api/service.go (hunk 11 above)
Comment (jtraglia) [reply services/api/service.go:1281]: Ah yeah nice catch!

**Review by michaelneuder — COMMENTED** [review]
tiny nits, still wrapping my head around it. will take another pass when im more fresh tomorrow

File: beaconclient/multi_beacon_client.go [hunk 15]
```diff
@@ -255,10 +259,20 @@ type publishResp struct {
...
+	blockHash, err := block.ExecutionBlockHash()
+	if err != nil {
+		c.log.WithError(err).Warn("failed to publish block as block hash is missing")
+		return 0, err
+	}
 	log := c.log.WithFields(logrus.Fields{
-		"slot":      block.Slot(),
-		"blockHash": block.BlockHash(),
+		"slot":      slot,
+		"blockHash": blockHash.String(),
```
[the remark this answers is not in the document]
Comment (avalonche) [reply beaconclient/multi_beacon_client.go:275]: why would it coerce into string? wouldn't it log [12, 23, 30 ...]?

File: beaconclient/prod_beacon_instance.go [hunk 16]
```diff
@@ -242,7 +242,7 @@ type GetBlockResponse struct {
 		Message struct {
 			Slot uint64 `json:"slot,string"`
 			Body struct {
-				ExecutionPayload types.ExecutionPayload `json:"execution_payload"`
+				ExecutionPayload capella.ExecutionPayload `json:"execution_payload"`
```
[the remark this answers is not in the document]
Comment (avalonche) [reply beaconclient/prod_beacon_instance.go:245]: No but this is not used anywhere, I will remove it

File: beaconclient/prod_beacon_instance.go (hunk 6 above)
Comment (avalonche) [reply beaconclient/prod_beacon_instance.go:277]: fetchBeacon() used to have different parameters but was changed due to ssz encoding issues with lighthouse. I'm planning to add a flag to toggle between json / ssz publishing to the beacon node in a future PR

File: common/test_utils.go [hunk 17]
```diff
@@ -64,23 +78,142 @@ var ValidPayloadRegisterValidator = boostTypes.SignedValidatorRegistration{
...
+				Version: spec.DataVersionDeneb,
+				Deneb: &builderApiDeneb.SubmitBlockRequest{
+					Message:   &bid.BidTrace,
+					Signature: [96]byte(signature),
+					ExecutionPayload: &deneb.ExecutionPayload{ //nolint:exhaustruct
+						Transactions:  []bellatrix.Transaction{[]byte{0x03}},
+						Timestamp:     bid.Slot * 12, // 12 seconds per slot.
+						PrevRandao:    _HexToHash("0xcf8e0d4e9587369b2301d0790347320302cc0943d5a1884560367e8208d920f2"),
+						Withdrawals:   []*capella.Withdrawal{},
+						BaseFeePerGas: uint256.NewInt(0),
```
[the remark this answers is not in the document]
Comment (avalonche) [reply common/test_utils.go:96]: Didn't think it was needed but good to add :thumbsup:

File: common/types.go [hunk 18]
```diff
@@ -1077,7 +598,7 @@ func (s *SubmitBlockRequestV2Optimistic) SizeSSZ() (size int) {
 
 	// Field (1) 'ExecutionPayloadHeader'
 	if s.ExecutionPayloadHeader == nil {
-		s.ExecutionPayloadHeader = new(consensuscapella.ExecutionPayloadHeader)
+		s.ExecutionPayloadHeader = new(capella.ExecutionPayloadHeader)
```
[the remark this answers is not in the document]
Comment (avalonche) [reply common/types.go:601]: to be honest I'm not too familiar with the optimistic code path and whether @michaelneuder would like to update this for deneb?
Otherwise I can try to add changes for the upgrade as a follow up PR

File: common/types_spec.go (hunk 3 above)
Comment (avalonche) [reply common/types_spec.go:247]: I think the builder's BuilderBlockValidationRequestV3 struct will unmashal to a zero value (0 hash) as it does not validate that it is a required field.

File: common/utils.go [hunk 19]
```diff
@@ -247,3 +180,119 @@ func GetEnvDurationSec(key string, defaultValueSec int) time.Duration {
...
+		Proposer:                   proposerPubkey,
+		ProposerFeeRecipient:       proposerFeeRecipient,
+		GasUsed:                    gasUsed,
+		GasLimit:                   gasLimit,
+		Timestamp:                  timestamp,
+		Transactions:               txs,
+		Value:                      value,
+		PrevRandao:                 prevRandao,
+		BlockNumber:                blockNumber,
+		Withdrawals:                withdrawals,
```
[the remark this answers is not in the document]
Comment (avalonche) [reply common/utils.go:275]: not needed as those fields are validated at the builder, this is just an helper struct that grabs the common fields needed for validation at the relay

File: datastore/memcached_test.go [hunk 20]
```diff
@@ -25,50 +35,89 @@ var (
...
+					Message: &builderApiV1.BidTrace{
+						Slot:                 1,
+						ParentHash:           phase0.Hash32{0x01},
+						BlockHash:            phase0.Hash32{0x09},
+						BuilderPubkey:        pubkey,
+						ProposerPubkey:       phase0.BLSPubKey{0x03},
+						ProposerFeeRecipient: bellatrix.ExecutionAddress{0x04},
+						Value:                uint256.NewInt(123),
+						GasLimit:             5002,
+						GasUsed:              5003,
```
[the remark this answers is not in the document]
Comment (avalonche) [reply datastore/memcached_test.go:55]: https://github.com/ethereum/builder-specs/blob/main/types/deneb/bid.yaml no just `blob_kzg_commitments` is new

File: beaconclient/multi_beacon_client.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (avalonche) [reply beaconclient/multi_beacon_client.go:265]: logrus.Fields is `type Fields map[string]interface{}` - I'm not sure if its formatted as a string here

File: common/types_spec.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (avalonche) [reply common/types_spec.go:256]: This is a limitation of having an unnamed field `*VersionedSubmitBlockRequest` and trying to do json marshalling in `BuilderBlockValidationRequest`. I can take a look if there are better ways of serialising this + tests for this so it is more robust.

File: beaconclient/multi_beacon_client.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (jtraglia) [reply beaconclient/multi_beacon_client.go:265]: Ah dang you're totally right. Sorry about that. It would print like this without the `.String()`:

```
blockHash="[1 2 3 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]"
```

**Comment by jtraglia** [comment]
@avalonche I didn't realize that switch statement pattern (`falllthrough`) was used to appease `exhaustive`.

**Comment by avalonche** [comment]
@jtraglia yeah I didn't like having to list every version as the code only supports a small subset of the versions, so disabled the linter

**Comment by metachris** [comment]
PSA: We've deployed https://github.com/flashbots/mev-boost-relay/commit/1ffdc1ad563ce7338f038a2d206d670acfc67197 to Flashbots mainnet relay.

**Review by metachris — APPROVED** [review]

**Review by metachris — APPROVED** [review]
